File: src/commands/dash.ts
import { existsSync, readdirSync, readFileSync, writeFileSync } from 'node:fs'
import { basename, join, relative, resolve } from 'node:path'
import { intro, log, outro, spinner } from '@clack/prompts'
import type { Command } from 'commander'
import ignore, { type Ignore } from 'ignore'
import color from 'picocolors'

// Em dash — (U+2014) and en dash – (U+2013)

const BINARY_EXTENSIONS = new Set([
  '.png',
  '.jpg',
  '.jpeg',
  '.gif',
  '.bmp',
  '.webp',
  '.ico',
  '.svg',
  '.mp3',
  '.mp4',
  '.wav',
  '.ogg',
  '.flac',
  '.avi',
  '.mov',
  '.mkv',
  '.zip',
  '.tar',
  '.gz',
  '.bz2',
  '.xz',
  '.rar',
  '.7z',
  '.pdf',
  '.doc',
  '.docx',
  '.xls',
  '.xlsx',
  '.ppt',
  '.pptx',
  '.exe',
  '.dll',
  '.so',
  '.dylib',
  '.bin',
  '.wasm',
  '.ttf',
  '.otf',
  '.woff',
  '.woff2',
  '.eot',
  '.db',
  '.sqlite',
  '.sqlite3',
  '.lock',
])

function isBinaryPath(filePath: string): boolean {
  const ext = filePath.slice(filePath.lastIndexOf('.')).toLowerCase()
  if (BINARY_EXTENSIONS.has(ext)) return true

  // Filenames without extension that are known binaries / non-text
  const base = basename(filePath).toLowerCase()
  const knownBinaryFiles = new Set([
    'pnpm-lock.yaml',
    'yarn.lock',
    'package-lock.json',
  ])
  if (knownBinaryFiles.has(base)) return false // still text, just skip for safety? No - these are text, keep them

  return false
}

function hasBinaryContent(buffer: Buffer): boolean {
  // Scan the first 8 KB for null bytes - a strong indicator of binary content
  const sampleSize = Math.min(buffer.length, 8192)
  for (let i = 0; i < sampleSize; i++) {
    if (buffer[i] === 0) return true
  }
  return false
}

function buildIgnore(cwd: string): Ignore {
  const ig = ignore()

  // Always ignore .git itself
  ig.add('.git')

  const gitignorePath = resolve(cwd, '.gitignore')
  if (existsSync(gitignorePath)) {
    const content = readFileSync(gitignorePath, 'utf-8')
    ig.add(content)
  }

  return ig
}

function walkFiles(
  dir: string,
  cwd: string,
  ig: Ignore,
  files: string[] = [],
  depth = 0,
  maxDepth = 50
): string[] {
  if (depth > maxDepth) return files
  let entries: import('node:fs').Dirent[]

  try {
    entries = readdirSync(dir, { withFileTypes: true })
  } catch {
    return files
  }

  for (const entry of entries) {
    const fullPath = join(dir, entry.name)
    const relPath = relative(cwd, fullPath).replace(/\\/g, '/')

    if (entry.isDirectory()) {
      // Check directory with and without trailing slash
      if (ig.ignores(relPath) || ig.ignores(`${relPath}/`)) continue
      walkFiles(fullPath, cwd, ig, files, depth + 1, maxDepth)
    } else if (entry.isFile()) {
      if (ig.ignores(relPath)) continue
      files.push(fullPath)
    }
  }

  return files
}

export function loadDashCommand(program: Command) {
  program
    .command('dash')
    .description(
      'Replace AI em dashes (— –) with a regular minus (-) in all non-ignored files'
    )
    .option('--dry-run', 'Preview changes without writing to disk', false)
    .action(async (options: { dryRun: boolean }) => {
      console.log()
      intro(color.inverse(' Opheys CLI - Dash Replacer '))

      const cwd = process.cwd()
      const ig = buildIgnore(cwd)

      const gitignorePath = resolve(cwd, '.gitignore')
      if (existsSync(gitignorePath)) {
        log.info(
          color.dim(`Using .gitignore from ${color.bold(gitignorePath)}`)
        )
      } else {
        log.warn(color.yellow('No .gitignore found - scanning all files'))
      }

      if (options.dryRun) {
        log.warn(color.yellow('Dry-run mode: no files will be modified'))
      }

      const s = spinner()
      s.start('Scanning files…')

      const allFiles = walkFiles(cwd, cwd, ig)

      s.stop(`Found ${color.bold(String(allFiles.length))} file(s) to inspect`)

      let scanned = 0
      let changed = 0
      let skipped = 0
      const changedFiles: string[] = []

      for (const filePath of allFiles) {
        scanned++

        if (isBinaryPath(filePath)) {
          skipped++
          continue
        }

        let buffer: Buffer
        try {
          buffer = readFileSync(filePath)
        } catch {
          skipped++
          continue
        }

        if (hasBinaryContent(buffer)) {
          skipped++
          continue
        }

        const original = buffer.toString('utf-8')

        if (!/[—–]/.test(original)) continue

        const replaced = original.replace(/[—–]/g, '-')
        const relPath = relative(cwd, filePath).replace(/\\/g, '/')

        if (!options.dryRun) {
          try {
            writeFileSync(filePath, replaced, 'utf-8')
          } catch {
            log.warn(color.yellow(`Could not write: ${relPath}`))
            continue
          }
        }

        changed++
        changedFiles.push(relPath)
        log.success(
          `${options.dryRun ? color.dim('[dry-run] ') : ''}${color.green(relPath)}`
        )
      }

      console.log()
      log.info(
        [
          `Scanned : ${color.bold(String(scanned))} file(s)`,
          `Modified: ${color.bold(color.green(String(changed)))} file(s)`,
          `Skipped : ${color.dim(String(skipped))} binary/unreadable file(s)`,
        ].join('\n          ')
      )

      if (changed === 0) {
        outro(color.dim('No em dashes found - nothing to do.'))
      } else if (options.dryRun) {
        outro(
          color.yellow(
            `Dry-run complete. ${changed} file(s) would have been modified.`
          )
        )
      } else {
        outro(color.green(`Done! Replaced em dashes in ${changed} file(s).`))
      }
    })
}

File: src/commands/dash.rs
use crate::ui;
use anyhow::Result;
use console::style;
use ignore::gitignore::{Gitignore, GitignoreBuilder};
use indicatif::{ProgressBar, ProgressStyle};
use std::fs;
use std::path::{Path, PathBuf};
use std::time::Duration;

const MAX_DEPTH: usize = 50;
const SAMPLE_SIZE: usize = 8192;

const BINARY_EXTENSIONS: &[&str] = &[
    ".png", ".jpg", ".jpeg", ".gif", ".bmp", ".webp", ".ico", ".svg", ".mp3", ".mp4", ".wav",
    ".ogg", ".flac", ".avi", ".mov", ".mkv", ".zip", ".tar", ".gz", ".bz2", ".xz", ".rar", ".7z",
    ".pdf", ".doc", ".docx", ".xls", ".xlsx", ".ppt", ".pptx", ".exe", ".dll", ".so", ".dylib",
    ".bin", ".wasm", ".ttf", ".otf", ".woff", ".woff2", ".eot", ".db", ".sqlite", ".sqlite3",
    ".lock",
];

fn build_ignore(cwd: &Path) -> Result<Gitignore> {
    let mut builder = GitignoreBuilder::new(cwd);
    builder.add_line(None, ".git")?;

    let gitignore_path = cwd.join(".gitignore");
    if gitignore_path.exists() {
        builder.add(gitignore_path);
    }

    Ok(builder.build()?)
}

fn is_binary_path(file_path: &Path) -> bool {
    let ext = file_path
        .extension()
        .and_then(|e| e.to_str())
        .map(|e| format!(".{e}"))
        .unwrap_or_default()
        .to_lowercase();
    BINARY_EXTENSIONS.contains(&ext.as_str())
}

fn has_binary_content(buffer: &[u8]) -> bool {
    let sample_size = buffer.len().min(SAMPLE_SIZE);
    buffer[..sample_size].contains(&0)
}

fn walk_files(dir: &Path, cwd: &Path, ig: &Gitignore, depth: usize) -> Vec<PathBuf> {
    if depth > MAX_DEPTH {
        return Vec::new();
    }

    let mut files = Vec::new();
    let entries = match fs::read_dir(dir) {
        Ok(e) => e,
        Err(_) => return files,
    };

    for entry in entries.flatten() {
        let full_path = entry.path();
        let rel_path = match full_path.strip_prefix(cwd) {
            Ok(p) => p,
            Err(_) => continue,
        };

        let rel_str = rel_path.to_string_lossy().replace('\\', "/");

        if entry.file_type().map(|t| t.is_dir()).unwrap_or(false) {
            if ig.matched(rel_path, true).is_ignore()
                || ig
                    .matched(Path::new(&format!("{rel_str}/")), true)
                    .is_ignore()
            {
                continue;
            }
            files.extend(walk_files(&full_path, cwd, ig, depth + 1));
        } else if entry.file_type().map(|t| t.is_file()).unwrap_or(false) {
            if ig.matched(rel_path, false).is_ignore() {
                continue;
            }
            files.push(full_path);
        }
    }

    files
}

fn contains_dashes(content: &str) -> bool {
    content.contains(['\u{2014}', '\u{2013}'])
}

fn replace_dashes(content: &str) -> String {
    content
        .chars()
        .map(|c| match c {
            '\u{2014}' | '\u{2013}' => '-',
            other => other,
        })
        .collect()
}

pub fn run(dry_run: bool) -> Result<()> {
    ui::intro(" oph - Dash Replacer ");

    let cwd = std::env::current_dir()?;
    let ig = build_ignore(&cwd)?;

    let gitignore_path = cwd.join(".gitignore");
    if gitignore_path.exists() {
        ui::log_info(&format!(
            "Using .gitignore from {}",
            style(gitignore_path.display()).bold()
        ));
    } else {
        ui::log_warn("No .gitignore found - scanning all files");
    }

    if dry_run {
        ui::log_warn("Dry-run mode: no files will be modified");
    }

    let pb = ProgressBar::new_spinner();
    pb.set_style(
        ProgressStyle::default_spinner()
            .template("{spinner} {msg}")
            .unwrap(),
    );
    pb.set_message("Scanning files…");
    pb.enable_steady_tick(Duration::from_millis(80));

    let all_files = walk_files(&cwd, &cwd, &ig, 0);
    pb.finish_and_clear();
    ui::log_info(&format!(
        "Found {} file(s) to inspect",
        style(all_files.len().to_string()).bold()
    ));

    let mut scanned = 0usize;
    let mut changed = 0usize;
    let mut skipped = 0usize;

    for file_path in &all_files {
        scanned += 1;

        if is_binary_path(file_path) {
            skipped += 1;
            continue;
        }

        let buffer = match fs::read(file_path) {
            Ok(b) => b,
            Err(_) => {
                skipped += 1;
                continue;
            }
        };

        if has_binary_content(&buffer) {
            skipped += 1;
            continue;
        }

        let original = match String::from_utf8(buffer) {
            Ok(s) => s,
            Err(_) => {
                skipped += 1;
                continue;
            }
        };

        if !contains_dashes(&original) {
            continue;
        }

        let replaced = replace_dashes(&original);
        let rel_path = file_path
            .strip_prefix(&cwd)
            .unwrap_or(file_path)
            .to_string_lossy()
            .replace('\\', "/");

        if !dry_run && fs::write(file_path, replaced).is_err() {
            ui::log_warn(&format!("Could not write: {rel_path}"));
            continue;
        }

        changed += 1;
        let prefix = if dry_run {
            format!("{} ", style("[dry-run]").dim())
        } else {
            String::new()
        };
        ui::log_success(&format!("{prefix}{}", style(rel_path).green()));
    }

    println!();
    ui::log_info(&format!(
        "Scanned : {} file(s)\n          Modified: {} file(s)\n          Skipped : {} binary/unreadable file(s)",
        style(scanned.to_string()).bold(),
        style(changed.to_string()).green().bold(),
        style(skipped.to_string()).dim()
    ));

    if changed == 0 {
        ui::outro("No em dashes found - nothing to do.");
    } else if dry_run {
        ui::outro(&format!(
            "Dry-run complete. {changed} file(s) would have been modified."
        ));
    } else {
        ui::outro(&format!("Done! Replaced em dashes in {changed} file(s)."));
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use tempfile::tempdir;

    #[test]
    fn replace_dashes_replaces_em_and_en() {
        let input = "foo-bar-baz";
        assert_eq!(replace_dashes(input), "foo-bar-baz");
    }

    #[test]
    fn respects_gitignore() -> Result<()> {
        let dir = tempdir()?;
        fs::write(dir.path().join(".gitignore"), "ignored/\n")?;
        fs::create_dir_all(dir.path().join("ignored"))?;
        fs::write(dir.path().join("ignored").join("file.txt"), "-")?;
        fs::write(dir.path().join("visible.txt"), "-")?;

        let ig = build_ignore(dir.path())?;
        let files = walk_files(dir.path(), dir.path(), &ig, 0);
        let names: Vec<_> = files
            .iter()
            .map(|p| p.file_name().unwrap().to_string_lossy().into_owned())
            .collect();

        assert!(names.contains(&"visible.txt".to_string()));
        assert!(!names.iter().any(|n| n == "file.txt"));
        Ok(())
    }

    #[test]
    fn dry_run_does_not_write() -> Result<()> {
        let dir = tempdir()?;
        let file = dir.path().join("test.txt");
        fs::write(&file, "hello-world")?;

        let content = fs::read_to_string(&file)?;
        assert!(contains_dashes(&content));
        let _ = replace_dashes(&content);
        assert!(fs::read_to_string(&file)?.contains('-'));
        Ok(())
    }
}
